apps/admin/src/lib/reports/send-reports.ts

```diff
@@ -1,12 +1,175 @@
-import { getUsersWitDueReport, updateLastReportTimestamp } from "@energyleaf/db/query";
+import { getElectricitySensorByUser } from "@/actions/sensors";
+import { env, getUrl } from "@/env.mjs";
+import {
+    createToken,
+    getEnergySumForSensorInRange,
+    getLastReportForUser,
+    getUserDataByUserId,
+    getUsersWitDueReport,
+    saveReport,
+    updateLastReportTimestamp,
+} from "@energyleaf/db/query";
+import { TokenType, type UserDataSelectType } from "@energyleaf/db/types";
+import { buildUnsubscribeReportsUrl } from "@energyleaf/lib";
+import type { DayStatistics, ReportProps } from "@energyleaf/lib";
+import { sendReport } from "@energyleaf/mail";
+
+interface UserReportData {
+    userId: string;
+    userName: string;
+    email: string;
+    receiveMails: boolean;
+    interval: number;
+}
 
 export async function createReportsAndSendMails() {
-    const userReportData = await getUsersWitDueReport();
+    const userReportData: Array<UserReportData> = await getUsersWitDueReport();
+
+    const totalReports = userReportData.length;
+    let successfulReports = 0;
+    let sentReports = 0;
+    let savedReports = 0;
+    let updatedLastReportTimestamps = 0;
+
+    for (const userReport of userReportData) {
+        let reportProps: ReportProps;
+        let unsubscribeLink = "";
+        try {
+            reportProps = await createReportData(userReport);
+            const unsubscribeToken = await createToken(userReport.userId, TokenType.Report);
+            unsubscribeLink = buildUnsubscribeReportsUrl({ baseUrl: getUrl(env), token: unsubscribeToken });
+        } catch (e) {
+            console.error(`Error creating report for User ${userReport.userName} (User-ID ${userReport.userId}): ${e}`);
+            continue;
+        }
+
+        if (userReport.receiveMails && reportProps) {
```

Comment:
`reportProps` kann in diesem Fall nicht undefined sein, da oben im Fehlerfall `continue` ausgeführt wird
```suggestion
        if (userReport.receiveMails) {
```